packages/docusaurus/src/server/plugins/init.ts

```diff
@@ -44,39 +45,127 @@ export default function initPlugins({
 
   const plugins: InitPlugin[] = pluginConfigs
     .map((pluginItem) => {
-      let pluginModuleImport: string | undefined;
+      // let pluginModuleImport: string | undefined;
       let pluginOptions: PluginOptions = {};
+      // let plugin: PluginModule | undefined;
+      let validateOptions;
+      // let pluginModule: ImportedPluginModule | undefined;
+      let pluginVersion: DocusaurusPluginVersionInformation;
 
       if (!pluginItem) {
         return null;
       }
 
-      if (typeof pluginItem === 'string') {
-        pluginModuleImport = pluginItem;
-      } else if (Array.isArray(pluginItem)) {
-        [pluginModuleImport, pluginOptions = {}] = pluginItem;
+      function getPlugin(): PluginModule | null {
+        if (typeof pluginItem === 'function') {
+          return pluginItem;
+        } else if (
+          Array.isArray(pluginItem) &&
+          typeof pluginItem[0] === 'function'
+        ) {
+          return pluginItem[0];
+        } else if (typeof pluginItem === 'string') {
+          const pluginModuleImport = pluginItem;
+          const pluginPath = pluginRequire.resolve(pluginModuleImport);
+          const pluginModule = importFresh<ImportedPluginModule>(pluginPath);
+          return pluginModule?.default || pluginModule;
+          // [path : string, options : {}]
+        } else if (
+          Array.isArray(pluginItem) &&
+          typeof pluginItem[0] === 'string'
+        ) {
+          const pluginModuleImport = pluginItem[0];
+          const pluginPath = pluginRequire.resolve(pluginModuleImport);
+          const pluginModule = importFresh<ImportedPluginModule>(pluginPath);
+          return pluginModule?.default || pluginModule;
+        }
+
+        return null;
+      }
+
+      function getPluginModuleImport(): string | null {
```

Comment:
Can you extract all those fns outside of initPlugins? I think it would help make the code more readable by passing the required data as function params (make the fn dependencies more explicit)

Reply:
The functions being inside a function where they are only used was common in docusaurus-plugin-content-docs so I thought I would follow that pattern since these functions are not used anywhere else

Reply:
The functions being inside a function where they are only used was common in docusaurus-plugin-content-docs so I thought I would follow that pattern since these functions are not used anywhere else

Reply:
This is not a big deal but splitting as much as possible into multiple pure functions makes the code more explicit, readable and testable.

Historically, most of the code of the docs plugin was directly inside index.test.ts, and  I moved code outside of it progressively.

Do your best, I'll continue the refactor later if you don't find a solution to extract those methods, it's not high priority and shouldn't block merging this PR.